NPF : User/group id filtering

Emmankoko opened the pull request.

Emmankoko committed:
Userland: npf rule parser for user and group id

Emmankoko committed:
Kernel: extract rule, lookup sockets ids, and process filtering

zoulasc commented on usr.sbin/npf/npfctl/npf_data.c:
```diff
@@ -267,6 +269,52 @@ npfctl_parse_table_id(const char *name)
+ * both uid and gid are both uint32_t
+ */
+struct r_id
+npfctl_init_rid(uint32_t id1, uint32_t id2, uint8_t op)
```
pass the rid pointer to be initialized to avoid extra copies.

Emmankoko committed:
pass rid_t by ref and copy the values directly to rid

Emmankoko committed:
npf_socket API

Emmankoko committed:
npfctl show

zoulasc commented on usr.sbin/npf/npfctl/npf_show.c:
```diff
@@ -548,6 +575,14 @@ npfctl_print_rule(npf_conf_info_t *ctx, nl_rule_t *rl, unsigned level)
 	}
 
+	if (!npf_rule_getrid(&rid, rl, "r_user")) {
+		ctx->fpos += fprintf(ctx->fp, "user \"%s\" ", print_guid(rid));
```
memory leaks

Emmankoko committed:
testing

in this testing, a socket is crafted in kernel to listen on any address( all interfaces) and port 65000 which is ideal for testing

to test,
run : npfctl debug -c $dir_to_npftest.conf -o ./npf.plist
npftest -c ./npf.plist -T guid -v

zoulasc commented on sys/net/npf/npf_handler.c:
```diff
@@ -251,7 +251,8 @@ npfk_packet_handler(npf_t *npf, struct mbuf **mp, ifnet_t *ifp, int di)
 	}
 
-	if (error) {
+	/* sidewaysh handling of rejecting packets whose addr-port pair matches no sockets  */
```
I think this should be done in a separate commit and "sideways"?

zoulasc commented on sys/net/npf/npf_ruleset.c:
```diff
@@ -1025,16 +1025,20 @@ npf_rule_match_rid(npf_rule_t *rl, npf_cache_t *npc, int dir)
 	}
 
-	return matched;
+	/* make sure if both uid and gid is set on rule, both must be mtching to agree */
```
"matching"

zoulasc commented on sys/net/npf/npf_ruleset.c:
```diff
@@ -1013,7 +1013,7 @@ int
 {
 	uint32_t sock_gid, sock_uid;
-	int matched = 0;
+	int uid_matched = 0, gid_matched = 0;
```
perhaps bool and false?

zoulasc commented on usr.sbin/npf/npftest/libnpftest/npf_rid_test.c:
```diff
@@ -0,0 +1,270 @@
+/*
```
Copyright, rcsid.

zoulasc commented on usr.sbin/npf/npftest/libnpftest/npf_rid_test.c:
```diff
@@ -0,0 +1,270 @@
+		int id_match;
+
+		id_match = npf_rule_match_rid(rl, npc, t->di);
+		error = npf_rule_conclude(rl, &mi);
```
error gets overwritten below and not tested here.

zoulasc commented on usr.sbin/npf/npftest/libnpftest/npf_rid_test.c:
```diff
@@ -0,0 +1,270 @@
+static struct socket *
+test_socket(int dir, uid_t uid, gid_t gid)
+{
+	struct sockaddr_in Server;
```
Why caps?

zoulasc commented on usr.sbin/npf/npftest/libnpftest/npf_rid_test.c:
```diff
@@ -0,0 +1,270 @@
+test_socket(int dir, uid_t uid, gid_t gid)
+{
+	struct sockaddr_in Server;
+	struct lwp *cur =  curlwp;
```
single space.

zoulasc commented on usr.sbin/npf/npftest/libnpftest/npf_rid_test.c:
```diff
@@ -0,0 +1,270 @@
+
+	struct socket *so;
+	int error = socreate(AF_INET, &so, SOCK_DGRAM, 0, cur, NULL);
+	if(error) {
```
Space after keyword (everywhere)

zoulasc replied:
why printf instead of warn(3)?

zoulasc commented on usr.sbin/npf/npftest/libnpftest/npf_rid_test.c:
```diff
@@ -0,0 +1,270 @@
+	sounlock(so);
+
+	if((error = sobind(so, (struct sockaddr *)&Server, cur)) != 0) {
+		printf("bind failed %d\n", error);
```
Again warn is better (everywhere)?

zoulasc commented on usr.sbin/npf/npftest/libnpftest/npf_rid_test.c:
```diff
@@ -0,0 +1,270 @@
+static bool
+test_static(bool verbose)
+{
+	for (unsigned i = 0; i < __arraycount(test_cases); i++) {
```
size_t is prolly better.

zoulasc reviewed (changes requested):
see inline comments.

Emmankoko committed:
docs
